### PR TITLE
[PLT-7608] Migrate Auditor Sign Certificates

### PR DESCRIPTION
**Pull request by amnambiar**
https://input-output.atlassian.net/browse/PLT-7608

BE PR - https://github.com/input-output-hk/dapps-certification/pull/90

@coderabbitai: ignore

### Summary by CodeRabbit

- New Feature: Introduced `CertificationMetadataForm` for managing certification metadata, including dynamic addition and removal of DApp scripts.
- New Feature: Updated `CreateCertificate` component to use the new `CertificationMetadata` component and improved its rendering logic.
- Refactor: Enhanced `DAPPScript` component with new props and state management for capturing deeply nested errors.
- New Feature: Added optional styling and behavior customization to the `Modal` component.
- Refactor: Improved validation and error messages in `ReportUpload` form and added a success modal after form submission.
- New Feature: Added `AuditorRunTestForm` component to replace the previous `CertificationForm` component in `Certification.tsx`.
- Refactor: Updated `UserProfile` interface and added new thunks for updating and fetching user profiles in `auth.slice.ts`.
- New Feature: Added a utility function to transform empty string values to `null` in an object.

### CODE REVIEW

File: src/pages/auditing/reportUpload/ReportUpload.tsx [hunk 1]
```diff
@@ -51,287 +33,134 @@ const ReportUpload = () => {
...
-        ?.some(
-          (field: { scriptHash: any; contractAddress: any }) =>
-            !field?.scriptHash || !field?.contractAddress
-        ); // prevent addition of new script boxes if the required field is empty
+        const timeout = setTimeout(() => {
+          clearTimeout(timeout);
+          setShowError("");
+        }, 5000);
+        setSubmitting(false);
+      });
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/auditing/reportUpload/ReportUpload.tsx]: ignoring

**Comment by RSoulatIOHK** [comment]
The metadata is not prefilled for me:

https://github.com/input-output-hk/dapps-certification-web/assets/117812549/6a7df07e-17bd-4796-9c1f-f0876ba7ee7e

This was done on:
BE: https://dapps-certification.scdev.aws.iohkdev.io/
FE: PR20 latest commit
DApp: [minimal-ptt-examples vesting latest commit](https://github.com/Ali-Hill/minimal-ptt-examples/commit/166f9b01c5083f3edbde0f7edf10e9b4af14c070)

The whole form behaves in a strange manner:

https://github.com/input-output-hk/dapps-certification-web/assets/117812549/ec62e000-8812-4d9e-873f-a1db63e04067

**Comment by amnambiar** [comment]
The issue was with commit  https://github.com/input-output-hk/dapps-certification-web/pull/20/commits/30150a4c7643858214b6132559f10b99f71f0c81 , to prefill form fiels (name, email, twitter, website) from profile fields.

Have reverted the commit to address this in a new bug [PLT-7662](https://input-output.atlassian.net/browse/PLT-7662).

File: src/pages/session/components/AppLayout.tsx [hunk 2]
```diff
@@ -29,6 +32,14 @@ const Banner = () => {
 }
 
 const AppLayout = () => {
+
+  const dispatch = useAppDispatch();
+  
+  useEffect(() => {
+    dispatch(fetchProfile({}))
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/session/components/AppLayout.tsx:41]: this is to be removed after merge, ignoring comment

File: src/pages/certification/components/AuditorRunTestForm/AuditorRunTestForm.tsx [hunk 3]
```diff
@@ -167,8 +168,8 @@ const AuditorRunTestForm: React.FC<IAuditorRunTestForm> = ({
             version: version,
             subject: subject
           },
-        });
-        if (profileResponse.data.dapp.owner) {
+        }));
+        if (response.payload && response.payload?.dapp?.owner) {
           const response = await postData.post("/run", checkout);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/components/AuditorRunTestForm/AuditorRunTestForm.tsx]: fixed

File: src/utils/utils.ts [hunk 4]
```diff
@@ -77,6 +77,15 @@ export const getObjectByPath = (object: { [x: string]: any }, path: string): any
...
 };
 
+export const transformEmptyStringToNullInObj = (obj: any) => {
+  Object.keys(obj).forEach(key => {
+    if (!obj[key]) {
+      obj[key] = null
+    } 
+  })
+  return obj;
+}
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/utils/utils.ts]: ignoring for now

File: src/components/CertificationMetadataForm/CertificationMetadataForm.tsx [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            })
+          }
+          {...props.form.register(props.config.id)}
+          {...rest}
+        />
+      );
+    default:
+      return null;
+  }
+};
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/components/CertificationMetadataForm/CertificationMetadataForm.tsx]: later

File: src/utils/utils.ts [hunk 6]
```diff
@@ -77,6 +77,15 @@ export const getObjectByPath = (object: { [x: string]: any }, path: string): any
...
+
 export const getErrorMessage = (errorObj: any) => {
   let errorMsg = "Something wrong occurred. Please try again later.";
     if (typeof errorObj === "string") {
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/utils/utils.ts]: ignoring

File: src/components/CertificationMetadataForm/CertificationMetadataForm.tsx [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    return (
+      !!form?.formState.errors?.[fieldArrayName] || // disable button if errors associated with dynamic form exists
+      form
+        .getValues(fieldArrayName)
+        ?.some(
+          (field: { scriptHash: any; contractAddress: any }) =>
+            !field?.scriptHash || !field?.contractAddress
+        )
+    ); // prevent addition of new script boxes if the required field is empty
+  }, [form]);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/components/CertificationMetadataForm/CertificationMetadataForm.tsx]: ignoring

File: src/components/CertificationMetadataForm/CertificationMetadataForm.tsx [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+  const initializeFormState = () => {
+    form.clearErrors(); // clear form errors
+    form.reset(initData);
+  };
+
+  useEffect(() => {
+    initializeFormState();
+    // initializeFormState() is to not to be triggered on every re-render of the dep-array below but whenever the form or userDetails is updated alone
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [form, initData]);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/components/CertificationMetadataForm/CertificationMetadataForm.tsx]: ignoring

File: src/pages/certification/components/CertificationMetadata/CertificationMetadata.tsx [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+      const { scriptHash, contractAddress, ...rest } = script;
+      return({
+        scriptHash: scriptHash,
+        ...(contractAddress && { contractAddress: contractAddress }),
+        smartContractInfo: {
+          ...transformEmptyStringToNullInObj(rest),
+        },
+      });
+    });
+  };
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/components/CertificationMetadata/CertificationMetadata.tsx]: ignoring

File: src/store/slices/auth.slice.ts [hunk 10]
```diff
@@ -172,7 +174,7 @@ export const startListenWalletChanges = createAsyncThunk('listenWalletChanges',
         await new Promise(resolve => setTimeout(resolve, 1000));
         isListening = (getState() as RootState).auth.listeningWalletChanges;
       }
-    } catch (error: any) {}
+    } catch (error: any) {throw new Error()}
   }
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/store/slices/auth.slice.ts:178]: ignoring here

File: src/components/CreateCertificate/CreateCertificate.tsx [hunk 11]
```diff
@@ -128,41 +155,83 @@ const CreateCertificate: React.FC<{ uuid: string; }> = ({ uuid }) => {
...
-    </>);
+                <Alert
+                severity="error" variant="filled"
+                onClose={() => setShowError("")}
+                >
+                    {showError}
+                </Alert>
+            </Snackbar>
+        </>
+    );
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/components/CreateCertificate/CreateCertificate.tsx]: not considered. ignoring

File: src/pages/auditing/reportUpload/ReportUpload.tsx [hunk 12]
```diff
@@ -51,287 +33,134 @@ const ReportUpload = () => {
...
-          ...rest
-        }
-      })
+          ...transformEmptyStringToNullInObj(rest),
+        },
+      });
     });
     return formattedScripts;
-  }
+  };
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/auditing/reportUpload/ReportUpload.tsx]: later

File: src/pages/auditing/reportUpload/ReportUpload.tsx [hunk 13]
```diff
@@ -51,287 +33,134 @@ const ReportUpload = () => {
...
-          (field: { scriptHash: any; contractAddress: any }) =>
-            !field?.scriptHash || !field?.contractAddress
-        ); // prevent addition of new script boxes if the required field is empty
+        const timeout = setTimeout(() => {
+          clearTimeout(timeout);
+          setShowError("");
+        }, 5000);
+        setSubmitting(false);
+      });
   };
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/auditing/reportUpload/ReportUpload.tsx]: later

File: src/pages/auditing/reportUpload/ReportUpload.tsx [hunk 14]
```diff
@@ -51,287 +33,134 @@ const ReportUpload = () => {
...
+        },
       },
-      report: reportURL.replace(/\s+/g,"").split(','),
+      report: reportURL.replace(/\s+/g, "").split(","),
       summary: summary,
       disclaimer: disclaimer,
-      scripts: formattedDappScripts
-    }
+      scripts: formattedDappScripts,
+    };
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/auditing/reportUpload/ReportUpload.tsx]: later

**Review by RSoulatIOHK — APPROVED** [review]
Approved following @amnambiar's comment

**Pull request merged**